src/plots/cartesian/set_convert.js

```diff
@@ -182,8 +182,9 @@ module.exports = function setConvert(ax) {
             // encounters them, ie all the categories from the
             // first data set, then all the ones from the second
             // that aren't in the first etc.
-            // TODO: sorting options - do the sorting
-            // progressively here as we insert?
+            // it is assumed that this function is being invoked in the
+            // already sorted category order; otherwise there would be
+            // a disconnect between the array and the index returned
```

Comment:
Filling in `ax._categories` within the [default step](https://github.com/plotly/plotly.js/blob/e03e9e0bd7ea3a6bed3b32bd784276e8e9eb70f1/src/plots/cartesian/axis_defaults.js) would definitely work (albeit with a slight performance hit).

We already need to loop all traces to check for box plot irregularities [here](https://github.com/plotly/plotly.js/blob/e03e9e0bd7ea3a6bed3b32bd784276e8e9eb70f1/src/plots/cartesian/axis_defaults.js#L156), so maybe you could fill in `ax._categories` within the same loop.